lib/handler/cache-handler.js

```diff
@@ -0,0 +1,355 @@
+        statusMessage,
+        headers
+      )
+    }
+
+    const now = Date.now()
+    const staleAt = determineStaleAt(headers, cacheControlDirectives)
+    if (staleAt) {
+      const varyDirectives = headers.vary
+        ? parseVaryHeader(headers.vary, this.#req.headers)
+        : undefined
+      const deleteAt = determineDeleteAt(cacheControlDirectives, staleAt)
+
+      const strippedHeaders = stripNecessaryHeaders(
+        rawHeaders,
+        headers,
+        cacheControlDirectives
+      )
+
+      this.#value = {
+        complete: false,
+        statusCode,
+        statusMessage,
+        rawHeaders: strippedHeaders,
+        body: [],
+        vary: varyDirectives,
+        size: currentSize,
+        cachedAt: now,
+        staleAt: now + staleAt,
+        deleteAt: now + deleteAt
+      }
+    }
+
+    return this.#handler.onHeaders(
+      statusCode,
+      rawHeaders,
+      resume,
+      statusMessage,
+      headers
+    )
+  }
+
+  /**
+   * @see {DispatchHandlers.onData}
+   *
+   * @param {Buffer} chunk
+   * @returns {boolean}
+   */
+  onData (chunk) {
+    if (this.#value) {
+      this.#value.size += chunk.length
+
+      if (this.#value.size > this.#getMaxEntrySize()) {
+        this.#value = null
+      } else {
+        this.#value.body.push(chunk)
+      }
+    }
+
+    return this.#handler.onData(chunk)
+  }
+
+  /**
+   * @see {DispatchHandlers.onComplete}
+   *
+   * @param {string[] | null} rawTrailers
+   */
+  onComplete (rawTrailers) {
+    if (this.#value) {
+      this.#value.complete = true
+      this.#value.rawTrailers = rawTrailers
+      this.#value.size += this.#getSizeOfBuffers(rawTrailers)
+
+      // If we're still under the max entry size, let's add it to the cache
+      if (this.#getMaxEntrySize() > this.#value.size) {
+        const result = this.#opts.store.put(this.#req, this.#value)
+        if (result && result.constructor.name === 'Promise') {
+          result.catch(err => this.#handler.onError(err))
+        }
+      }
+    }
+
+    return this.#handler.onComplete(rawTrailers)
+  }
+
+  /**
+   * @see {DispatchHandlers.onError}
+   *
+   * @param {Error} err
+   */
+  onError (err) {
+    this.#value = undefined
+    this.#handler.onError(err)
+  }
+
+  /**
+   * @returns {number}
+   */
+  #getMaxEntrySize () {
+    return this.#opts.store.maxEntrySize ?? Infinity
+  }
+
+  /**
+   * @param {string[] | Buffer[]} arr
+   * @returns {number}
+   */
+  #getSizeOfBuffers (arr) {
+    let size = 0
+
+    for (const buffer of arr) {
+      size += buffer.length
+    }
+
+    return size
+  }
+}
+
+/**
+ * @see https://www.rfc-editor.org/rfc/rfc9111.html#name-storing-responses-to-authen
+ *
+ * @param {number} statusCode
+ * @param {Record<string, string>} headers
+ * @param {import('../util/cache.js').CacheControlDirectives} cacheControlDirectives
+ */
+function canCacheResponse (statusCode, headers, cacheControlDirectives) {
+  if (![200, 307].includes(statusCode)) {
```

Comment:
avoid the function call and intermediary array construction.

```suggestion
  if (
    statusCode !== 200 &&
    statusCode !== 307
  ) {
```